mozldap.go

```diff
@@ -236,117 +221,113 @@ func (cli *Client) Search(base, filter string, attributes []string) (entries []l
+		return "", fmt.Errorf("%s -> found %d entries matching mail %q, expected 1", errorText, len(entries), email)
 	}
 	fullName = entries[0].GetAttributeValue("cn")
-	return
+	return fullName, err
+}
+
+// GetUserDNById exists for API compatiability (use GetUserDNByUID)
+func (cli *Client) GetUserDNById(uid string) (string, error) {
+	return cli.GetUserDNByUID(uid)
 }
 
-// GetUserDNByID returns the distinguished name of a given user using his ID
+// GetUserDNByUID returns the distinguished name of a given user using his ID
 //
-// example: cli.GetUserDNByID("jvehent")
-func (cli *Client) GetUserDNById(uid string) (dn string, err error) {
-	defer func() {
-		if e := recover(); e != nil {
-			err = fmt.Errorf("mozldap.GetUserDNByID(uid=%q) -> %v", uid, e)
-		}
-	}()
+// example: cli.GetUserDNByUID("jvehent")
+func (cli *Client) GetUserDNByUID(uid string) (string, error) {
+	errorText := fmt.Sprintf("mozldap.GetUserDNByUID(uid=%q)", uid)
+	dn := ""
+
 	entries, err := cli.Search("", "(uid="+uid+")", []string{"mail"})
 	if err != nil {
-		panic(err)
+		return "", fmt.Errorf("%s -> %q", errorText, err.Error())
 	}
 	if len(entries) != 1 {
-		panic(fmt.Sprintf("found %d entries matching uid %q, expected 1", len(entries), uid))
+		return "", fmt.Errorf("%s -> found %d entries matching uid %q, expected 1", errorText, len(entries), uid)
 	}
 	dn = entries[0].DN
-	return
+	return dn, nil
 }
 
-// GetUserUidNumber returns the UID number of a user using a shortdn
+// GetUserUidNumber exists for API compatiability (use GetUserUIDNumber)
+func (cli *Client) GetUserUidNumber(shortdn string) (uint64, error) {
+	return cli.GetUserUIDNumber(shortdn)
+}
+
+// GetUserUIDNumber returns the UID number of a user using a shortdn
 //
-// example: cli.GetUserUidNumber("mail=jvehent@mozilla.com")
-func (cli *Client) GetUserUidNumber(shortdn string) (uidNumber uint64, err error) {
-	defer func() {
-		if e := recover(); e != nil {
-			err = fmt.Errorf("mozldap.GetUserUidNumber(shortdn=%q) -> %v",
-				shortdn, e)
-		}
-	}()
+// example: cli.GetUserUIDNumber("mail=jvehent@mozilla.com")
+func (cli *Client) GetUserUIDNumber(shortdn string) (uint64, error) {
+	errorText := fmt.Sprintf("mozldap.GetUserUIDNumber(shortdn=%q)", shortdn)
+	uidNumber := uint64(0)
+
 	entries, err := cli.Search("", "("+shortdn+")", []string{"uidNumber"})
 	if err != nil {
-		panic(err)
+		return 0, fmt.Errorf("%s -> %q", errorText, err.Error())
 	}
-	for _, entry := range entries {
-		for _, attr := range entry.Attributes {
-			if attr.Name != "uidNumber" {
-				continue
-			}
-			for _, val := range attr.Values {
-				uidNumber, err = strconv.ParseUint(val, 10, 64)
-				if err != nil {
-					panic(err)
-				}
-			}
+
+	if entries[0].Attributes[0] != nil && len(entries[0].Attributes[0].Values) == 1 {
+		uidNumber, err = strconv.ParseUint(entries[0].GetAttributeValue("uidNumber"), 10, 64)
+		if err != nil {
+			return 0, fmt.Errorf("%s -> %q", errorText, err.Error())
```

Comment:
This is a change in logic here. Your version will probably work fine but is also less reliable. Are you sure about this?